### PR TITLE
protobuf 3.0 parser tool used for TensorFlow weights and graph

### PR DESCRIPTION
**Pull request by Wangyida**
Enough for parsing weights and graph from TensorFlow. Especially using GraphDef Class.

### CODE REVIEW

**Comment by nyanp** [comment]
@Wangyida 
Great :) Could you tell me your designing plan for importing TF? My concerns are:

- What we can do, and what we can't do by this PR
    - Constructing tiny-cnn graph networks from trained TF model?
    - Or just adding low-level functions for further plans?
- Prerequisites to run your codes
    - Target OS(windows/linux/ios/android?) / Build system(bazel?) / Other languages(python/numpy?)

**Comment by Wangyida** [comment]
@nyanp Maybe a little confusing about this TF related PR. Here I just provide a clean parser for TensorFlow trained graph and weights using protobuf 3.0, so there won't be any TensorFlow related building issues. This is a parser like caffe_converter.

For TensorFlow related implementation, I open a branch [tf](https://github.com/Wangyida/tiny-cnn/tree/tf). The advantage is that this building process of TF just using make for building rather than bazel. But from discussion with @bhack , this implementation just introduce TF into tiny-cnn with API of TF which is not so necessary and tiny anymore. I didn't open a PR on that branch.

**Comment by nyanp** [comment]
@Wangyida 
Ok, this PR is enough for mainstream of tiny :+1:  We need a doc which explain at least them

- Prerequisites 
- Concrete steps to construct nets

**Comment by Wangyida** [comment]
@nyanp 
Alright, now I use ```protoc``` to compile the generated source and header of from .proto files of tensorflow, these could all be included in the CMakeLists.txt of examples.
As for networking training script, this is an option. This parser cloud also load official train inception model from TensorFlow.

File: examples/tensorflow_converter/squeezenet_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Timing benchmark for SqueezeNet inference.
+
+To run, use:
+  bazel run -c opt --config=cuda \
+      third_party_tensorflow_models_image_squeezenet:squeezenet_benchmark
```
Comment (patrikhuber) [inline examples/tensorflow_converter/squeezenet_benchmark.py]: I didn't build this but on alexnet you use slashes ('/') and here it's underscores ('_') - one of them could be wrong?

Comment (Wangyida) [inline examples/tensorflow_converter/squeezenet_benchmark.py]: Thanks for notification, underscores is wrong. I use batch replacement on some operation name from original caffe prototxt for SqueezeNet which causes such mistake.

**Comment by edgarriba** [comment]
@Wangyida 
I vote to move `examples/tensorflow_converter/*.h` files to `tiny_cnn/io`

**Comment by Wangyida** [comment]
@edgarriba OK, seems right.

File: examples/tensorflow_converter/example_tensorflow_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+template<class in_value>
+void all2str(string & result, const in_value& t)
+{
+    ostringstream oss;
+    oss<<t;
+    result=oss.str();
+}
+
+/* This method is used as reference for bit operation for char* parsering
```
Comment (edgarriba) [inline examples/tensorflow_converter/example_tensorflow_parser.cpp]: if it's not needed, remove it?

File: examples/tensorflow_converter/alexnet_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Comment (edgarriba) [inline examples/tensorflow_converter/alexnet_benchmark.py]: do we really need this python files?

File: examples/tensorflow_converter/node_factory_impl.h [hunk 4]
```diff
@@ -0,0 +1,1156 @@
...
+
+        if (bn_param.has_eps()) {
+            eps = bn_param.eps();
+        }
+        if (bn_param.has_moving_average_fraction()) {
+            momentum = bn_param.moving_average_fraction();
+        }
+    }
+
+    auto bn = std::make_shared<bn_layer>(bottom_shape.area(), bottom_shape.depth_, eps, momentum, net_phase::test);
```
Comment (edgarriba) [inline examples/tensorflow_converter/node_factory_impl.h]: format this

File: examples/tensorflow_converter/node_factory_impl.h [hunk 5]
```diff
@@ -0,0 +1,1156 @@
...
+                                        layer *dst) {
+    auto weights = src.blobs(0);
+    int curr = 0;
+
+    if (dst->out_size() * dst->in_size() !=
+        static_cast<cnn_size_t>(weights.data_size())) {
+        throw nn_error(
+            std::string("layer size mismatch!") +
+            "caffe(" + src.name() + "):" + to_string(weights.data_size()) + "\n" +
+            "tiny-cnn(" + dst->layer_type() + "):" + to_string(dst->get_weights().size()));
```
Comment (edgarriba) [inline examples/tensorflow_converter/node_factory_impl.h]: tiny-dnn

File: examples/tensorflow_converter/node_factory_impl.h [hunk 6]
```diff
@@ -0,0 +1,1156 @@
...
+
+        vec_t& w = *dst->get_weights()[0];
+        vec_t& b = *dst->get_weights()[1];
+
+        if (!w.empty()) {
+            std::fill(w.begin(), w.end(), weight);
+        }
+        if (!b.empty()) {
+            std::fill(b.begin(), b.end(), float_t(0));
+            //dst->init_bias();
```
Comment (edgarriba) [inline examples/tensorflow_converter/node_factory_impl.h]: remove it if it's not needed

File: examples/tensorflow_converter/node_factory_impl.h [hunk 7]
```diff
@@ -0,0 +1,1156 @@
...
+                    if (node_table[input]) {
+                        auto& bottom = node_table[input];
+                        n.prev.push_back(bottom);
+                        // layer_table[bottom->layer->name()]->next = &n;
+                    }
+                    /*
+                    if (n.layer->top_size() > 0) {
+                        node_table[n.layer->top(0)] = &n;
+                    }
+                    */
```
Comment (edgarriba) [inline examples/tensorflow_converter/node_factory_impl.h]: remove this

File: examples/tensorflow_converter/protoc.sh [hunk 8]
```diff
@@ -0,0 +1 @@
+protoc -I=..  --cpp_out=.. ../tensorflow/core/framework/*.proto
```
Comment (edgarriba) [inline examples/tensorflow_converter/protoc.sh]: ??

File: examples/tensorflow_converter/squeezenet_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,580 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Comment (edgarriba) [inline examples/tensorflow_converter/squeezenet_benchmark.py]: Does tiny-dnn need this file?

File: examples/tensorflow_converter/protoc.sh (hunk 8 above)
Comment (Wangyida) [inline examples/tensorflow_converter/protoc.sh]: This shell script is used for other header and source generating apart from the main .proto file of TensorFlow file. If we include all compiled headers and source files, there will be a lot of useless files which might still be useful for further implementation.

File: examples/tensorflow_converter/alexnet_benchmark.py (hunk 3 above)
Comment (Wangyida) [inline examples/tensorflow_converter/alexnet_benchmark.py]: I removed these useless files which are just used for training the graph.

File: examples/tensorflow_converter/example_tensorflow_parser.cpp (hunk 2 above)
Comment (Wangyida) [inline examples/tensorflow_converter/example_tensorflow_parser.cpp]: This is used for attribution extraction.

File: examples/tensorflow_converter/node_factory_impl.h (hunk 6 above)
Comment (Wangyida) [inline examples/tensorflow_converter/node_factory_impl.h]: I removed some un-mature files for graph loader.

File: examples/tensorflow_converter/example_tensorflow_parser.cpp (hunk 2 above)
Comment (edgarriba) [inline examples/tensorflow_converter/example_tensorflow_parser.cpp]: I mean methods below static `int32_t BigEndInt` and `static float BigEndFloat`

Comment (Wangyida) [inline examples/tensorflow_converter/example_tensorflow_parser.cpp]: Yeah I see, already removed. The IOS codes might be pushed tommorow as a standalone PR~

**Comment by edgarriba** [comment]
@Wangyida 
Are you planning to provide a minimum API ?
I think that the routines in `examples/tensorflow_converter/example_tensorflow_parser.cpp ` should be placed into `tiny_dnn/io/tensorflow_converter` in form of an API.

/cc @bhack @nyanp

**Comment by Wangyida** [comment]
@edgarriba You are right, I will recollect these files in the io folder to be as an API.

**Comment by Wangyida** [comment]
@edgarriba OK, already as you suggested.

File: tiny_cnn/io/tensorflow/proto_parser.h [hunk 10]
```diff
@@ -0,0 +1,256 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <map>
+#include "tensorflow/core/framework/graph.pb.h"
+using namespace std;
```
Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
Please don't use ```using namespace``` in header files. It will affect user codes. 
http://stackoverflow.com/questions/5849457/using-namespace-in-c-headers

File: tiny_cnn/io/tensorflow/proto_parser.h [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+using namespace tensorflow;
+
+template<class in_value>
```
Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
We need to move all functions under ```namespace tiny_cnn``` (```tiny_dnn``` for master branch)

File: tiny_cnn/io/tensorflow/proto_parser.h [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+void all2str(string & result, const in_value& t)
+{
+    ostringstream oss;
+    oss<<t;
+    result=oss.str();
+}
+
+static float char_to_float(const char* b) {
```
Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
use ```inline``` for all functions in this file.

File: tiny_cnn/io/tensorflow/proto_parser.h [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+// Iterates though all nodes in the GraphDef and prints info about them.
+void ListNodes(const tensorflow::GraphDef& graph_def) {
+  cout << "  Node Size: " << graph_def.node_size() << endl;
+  for (int i = 0; i < graph_def.node_size(); i++) {
+    const tensorflow::NodeDef& node_def = graph_def.node(i);
+    // print nodes information
+    fprintf(stderr, "\nNode: %s (%s)\n", node_def.name().c_str(), node_def.op().c_str());
+    fprintf(stderr, "  Is Placeholder: %d, Is Const: %d\n", IsPlaceholder(node_def), IsScalarConst(node_def));
```
Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
Is there any reason to use ```stderr```? If tiny-dnn use ```stderr``` even when we don't have any problem, it would confuse 
 users very much.

File: tiny_cnn/io/tensorflow/proto_parser.h (hunk 12 above)
Comment (bhack) [inline tiny_cnn/io/tensorflow/proto_parser.h]: I use this ad reference https://isocpp.org/wiki/faq/inline-functions#inline-and-perf

File: tiny_cnn/io/tensorflow/proto_parser.h [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+  return f;
+}
+
+static float char_to_int32(const char* b) {
+  int32_t i;
+  memcpy(&i, b, sizeof(int32_t));
+  return i;
+}
+
+void summarize_attr_value(const string& attr_string, const AttrValue& attr_value) {
```
Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
We will violate the one definition rule (ODR) if we declare function without inline/static. see:
http://en.cppreference.com/w/cpp/language/definition

File: tiny_cnn/io/tensorflow/proto_parser.h (hunk 13 above)
Comment (Wangyida) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @nyanp I checked all your suggestions, I will modify them as desired.

File: tiny_cnn/io/tensorflow/proto_parser.h (hunk 14 above)
Comment (Wangyida) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @nyanp Very valuable for me, could have a check now for other improper definition?

Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: @Wangyida 
Good! I've done and it's ok :)

File: tiny_cnn/io/tensorflow/proto_parser.h (hunk 11 above)
Comment (pansk) [inline tiny_cnn/io/tensorflow/proto_parser.h]: Moreover, "using namespace" top-level in an include file is really a bad idea.

Comment (nyanp) [inline tiny_cnn/io/tensorflow/proto_parser.h]: Exactly, and @Wangyida already fixed this "using namespace" issue :)

**Comment by bhack** [comment]
Where is all this code in master?

**Comment by edgarriba** [comment]
Surprisingly it wasn't sync . @nyanp Can you merge these files?

**Comment by bhack** [comment]
Probably we could open also an issue with a list f TF tasks check points that need to be completed. @Wangyida Can you support us to open this ticket with a task list?

**Comment by edgarriba** [comment]
we definitely need to merge this to the master so that community can iterate over it

**Comment by nyanp** [comment]
Ok I'll merge this into master.
